Create GraphQL schema and generate interfaces

rning opened the pull request:
Created the GraphQL schema to be the Spartify API. This allows for development
to continue in parallel for the backend and frontend. Note that the backend
SQL table database models as specified during our Week 2 lab section should be
consistent with implementing this API.

Ran "npm run gen" to validate the schema and generate the TypeScript interfaces
for the GraphQL types.
https://github.com/rothfels/bespin/wiki/How-To-Guide

This is a breaking change, as the rest of the project still implements the class
website project.

Closes #2 
Closes #6

rning committed:
Create GraphQL schema and generate interfaces

Created the GraphQL schema to be the Spartify API. This allows for development
to continue in parallel for the backend and frontend. Note that the backend
SQL table database models as specified during our Week 2 lab section should be
consistent with implementing this API.

Ran "npm run gen" to validate the schema and generate the TypeScript interfaces
for the GraphQL types.
https://github.com/rothfels/bespin/wiki/How-To-Guide

This is a breaking change, as the rest of the project still implements the class
website project.

Rohanator9000 commented on server/src/graphql/schema.graphql:
```diff
@@ -11,81 +11,98 @@
-type Survey {
-  id: Int!
+  # The time that the Party was last interacted with. ISO 8601 format for date and time in UTC.
+  latestTime: String!
```
Why do we need to expose this variable? I can't think of a client-side use case.

rning replied:
Will our background process be interacting directly with the database layer, or through the API? If the former, then yeah I suppose we won't need latestTime.

Rohanator9000 replied:
Good question, I'm actually not sure. We can ask someone on Friday and remove this if its unneeded.
I don't know if we have to worry about security for this project, but it would be pretty insecure to allow unauthenticated API requests to delete parties.

Rohanator9000 commented on server/src/graphql/schema.types.ts:
```diff
@@ -12,82 +12,71 @@ export interface Scalars {
-  currentQuestion?: Maybe<SurveyQuestion>
-  questions: Array<Maybe<SurveyQuestion>>
+  password?: Maybe<Scalars['String']>
+  latestTime: Scalars['String']
```
Do you think it would be easier to keep this as an int (unix epoch time)? It might be easier to find the difference with the current time (for the background process), and I can't think of a reason it would need to be a string.

rning replied:
I chose a string representation just by habit since gql uses 32 bit ints, so I think we could only represent up to around the year 2038. It doesn't really matter for our project, so if everyone want to change the type to an int for our API then I'm good with that, too.

Rohanator9000 replied:
Oh ok that makes sense. I'm fine with leaving it as a string.

danielskathyd commented:
Do we need to write a mutation that gets called when a song finishes playing to add it to a party's played song list with an incremented sequence number? Also how should we determine when a song is finished playing to continue to the next one if we aren't technically playing it... Should we possibly add a time field to each song?

Rohanator9000 commented:
> Do we need to write a mutation that gets called when a song finishes playing to add it to a party's played song list with an incremented sequence number? Also how should we determine when a song is finished playing to continue to the next one if we aren't technically playing it... Should we possibly add a time field to each song?

He wrote a mutation called `nextSong(partyId: Int!): Party`. And iirc we're going to rely on a 'Next Song' button for now, but we could change it in the future.

danielskathyd commented on server/src/graphql/schema.graphql:
```diff
@@ -11,81 +11,98 @@
+  # Advances the CurrentSong of a Party to the next song, which is the VotedSong with the highest VotedSong.count
+  # out of all the VotedSong for the Party. If there is no VotedSong for the Party, the CurrentSong becomes null.
+  # Returns the Party if successful, or null if unsuccessful.
+  nextSong(partyId: Int!): Party
```
Should we add another comment specifying that the skipped song will be added to `playedSongs`

rning replied:
Sure, I'll do that once we agree on all the changes to make.

rning committed:
Add comment describing the nextSong mutation

This comment clarifies that the API's nextSong mutation causes the currentSong
of the Party prior to calling nextSong to be used to create a PlayedSong for
that Party.
